electrum/plugins/swapserver/server.py

```diff
@@ -0,0 +1,128 @@
+            "warnings": [],
+            "pairs": {
+                "BTC/BTC": {
+                    "hash": "dfe692a026d6964601bfd79703611af333d1d5aa49ef5fedd288f5a620fced60",
+                    "rate": 1,
+                    "limits": {
+                        "maximal": 10000000,
+                        "minimal": 28560,
+                        "maximalZeroConf": {
+                            "baseAsset": 0,
+                            "quoteAsset": 0
+                        }
+                    },
+                    "fees": {
+                        "percentage": 0.5,
+                        "minerFees": {
+                            "baseAsset": {
+                                "normal": 27880,
+                                "reverse": {
+                                    "claim": 22632,
+                                    "lockup": 25092
+                                }
+                            },
+                            "quoteAsset": {
+                                "normal": 27880,
+                                "reverse": {
+                                    "claim": 22632,
+                                    "lockup": 25092
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        return web.json_response(pairs)
+
+    async def create_swap(self, r):
+        sm = self.wallet.lnworker.swap_manager
+        await sm.get_pairs() # fixme
+        request = await r.json()
+        req_type = request['type']
+        assert request['pairId'] == 'BTC/BTC'
+        if req_type == 'reversesubmarine':
+            swap, payment_hash, invoice = sm.add_server_swap(
+                lightning_amount_sat=request['invoiceAmount'],
+                payment_hash=bytes.fromhex(request['preimageHash']),
+                their_pubkey=bytes.fromhex(request['claimPublicKey'])
+            )
+            response = {
+                'id': payment_hash.hex(),
+                'invoice': invoice,
+                'minerFeeInvoice': None,
+                'lockupAddress': swap.lockup_address,
+                'redeemScript': swap.redeem_script.hex(),
+                'timeoutBlockHeight': swap.locktime,
+                "onchainAmount": swap.onchain_amount,
+            }
+        elif req_type == 'submarine':
+            swap, payment_hash, invoice = sm.add_server_swap(
+                invoice=request['invoice'],
+                their_pubkey=request['refundPublicKey']
+            )
+            response = {
+                "id": payment_hash.hex(),
+                "acceptZeroConf": False,
+                "expectedAmount": swap.onchain_amount,
+                "timeoutBlockHeight": swap.locktime,
+                "address": swap.lockup_address,
+                "redeemScript": swap.redeem_script.hex()
+            }
+        else:
+            raise Exception('unsupported request type:' + req_type)
```

Comment:
what if `req_type` is not even a string? In general `r` is untrusted input.

1. I haven't checked what happens if this method raises -- but obviously that should not kill the swap server.
2. I would be more comfortable if all fields of the request would be validated as early as logically feasible. E.g. what happens if `claimPublicKey` is 1000 bytes long?